testutil/state/suite.go

```diff
@@ -0,0 +1,104 @@
+package state
+
+import (
+	"testing"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	"github.com/cosmos/cosmos-sdk/store"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/bank"
+	"github.com/stretchr/testify/require"
+	abci "github.com/tendermint/tendermint/abci/types"
+	dbm "github.com/tendermint/tm-db"
+
+	"github.com/ovrclk/akash/testutil"
+	dkeeper "github.com/ovrclk/akash/x/deployment/keeper"
+	dtypes "github.com/ovrclk/akash/x/deployment/types"
+	"github.com/ovrclk/akash/x/market/keeper"
+	"github.com/ovrclk/akash/x/market/types"
+	pkeeper "github.com/ovrclk/akash/x/provider/keeper"
+	ptypes "github.com/ovrclk/akash/x/provider/types"
+)
+
+// TestSuite encapsulates a functional Akash nodes data stores for
+// ephemeral testing.
+type TestSuite struct {
+	t       testing.TB
+	ms      sdk.CommitMultiStore
+	ctx     sdk.Context
+	mkeeper keeper.Keeper
+	dkeeper dkeeper.Keeper
+	pkeeper pkeeper.Keeper
+	bkeeper bank.Keeper
+}
+
+// T returns the testing.T struct of the suite.
+func (ts *TestSuite) T() *testing.T {
+	tFace := ts.t
+	var tStruct *testing.T = tFace.(*testing.T)
+	return tStruct
+}
+
+// SetupTestSuite provides toolkit for accessing stores and keepers
+// for complex data interactions.
+func SetupTestSuite(t testing.TB, codec *codec.Codec) *TestSuite {
+	suite := &TestSuite{
+		t: t,
+	}
+
+	mKey := sdk.NewKVStoreKey(types.StoreKey)
+	dKey := sdk.NewKVStoreKey(dtypes.StoreKey)
+	pKey := sdk.NewKVStoreKey(ptypes.StoreKey)
+
+	db := dbm.NewMemDB()
+	suite.ms = store.NewCommitMultiStore(db)
+	suite.ms.MountStoreWithDB(mKey, sdk.StoreTypeIAVL, db)
+	suite.ms.MountStoreWithDB(dKey, sdk.StoreTypeIAVL, db)
+	suite.ms.MountStoreWithDB(pKey, sdk.StoreTypeIAVL, db)
+
+	ts := suite.T()
+	err := suite.ms.LoadLatestVersion()
+	require.NoError(ts, err)
+	suite.ctx = sdk.NewContext(suite.ms, abci.Header{}, true, testutil.Logger(ts))
+
+	suite.mkeeper = keeper.NewKeeper(codec, mKey)
+	suite.dkeeper = dkeeper.NewKeeper(codec, dKey)
+	suite.pkeeper = pkeeper.NewKeeper(codec, pKey)
+
+	return suite
+}
+
+// SetBlockHeight provides arbitrarily setting the chain's block height.
+func (ts *TestSuite) SetBlockHeight(height int64) {
+	ts.ctx = ts.ctx.WithBlockHeight(height)
+}
+
+// Store provides access to the underlying KVStore
+func (ts *TestSuite) Store() sdk.CommitMultiStore {
+	return ts.ms
+}
+
+// Context of the current mempool
+func (ts *TestSuite) Context() sdk.Context {
+	return ts.ctx
+}
+
+// MarketKeeper key store
+func (ts *TestSuite) MarketKeeper() keeper.Keeper {
+	return ts.mkeeper
+}
+
+// DeploymentKeeper key store
+func (ts *TestSuite) DeploymentKeeper() dkeeper.Keeper {
+	return ts.dkeeper
+}
+
+// ProviderKeeper key store
+func (ts *TestSuite) ProviderKeeper() pkeeper.Keeper {
+	return ts.pkeeper
+}
+
+// BankKeeper key store
+func (ts *TestSuite) BankKeeper() bank.Keeper {
+	return ts.bkeeper
+}
```

Comment:
:+1: 

can it be re-used in existing tests?

Reply:
For the most part yes! However in the case of tests within the `handler` package, it can't because of the `app` package import. That causes an import cycle.

So it can replace testing of package's public API, but for internal methods(such as `matchOrders()` directly) it has to be a local copy-pasta.